rust/datafusion/src/sql/planner.rs

```diff
@@ -322,10 +317,80 @@ impl<'a, S: SchemaProvider> SqlToRel<'a, S> {
             ));
         }
 
-        let plan = self.plan_tables_with_joins(&select.from)?;
+        let plans = self.plan_from_tables(&select.from)?;
 
-        // filter (also known as selection) first
-        let plan = self.filter(&plan, &select.selection)?;
+        let plan = match &select.selection {
+            Some(predicate_expr) => {
+                // build join schema
+                let mut fields = vec![];
+                for plan in &plans {
+                    let schema = plan.schema();
+                    for field in schema.fields() {
+                        fields.push(field.clone());
+                    }
+                }
+                let join_schema = Schema::new(fields);
```

Comment:
I am not sure how this code handles duplicated column names in the two inputs. TPCH helpfully prefixes id columns with the relation (e.g. `l_orderkey` and `o_orderkey`) but in many real world inputs both tables would probably have a column named `orderkey`)

Reply:
Thanks @alamb. I will add the check for duplicate column names and merge this once the predicate push-down bug is resolved.

After this, I plan on working on table/relation aliases and qualified column names in queries [1].

https://issues.apache.org/jira/browse/ARROW-10732

Reply:
On second thoughts, I have ignored the failing test and added a link to the bug [1].

I have also added the check for duplicate column names after a join.

[1] https://issues.apache.org/jira/browse/ARROW-10760